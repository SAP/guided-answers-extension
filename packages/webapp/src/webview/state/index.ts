export { actions, store } from './store';
export { getInitialState } from './reducers';
