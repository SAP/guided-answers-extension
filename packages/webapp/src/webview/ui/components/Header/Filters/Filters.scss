.dialog-filter .ms-Dialog-content {
    overflow: hidden;
}

ul.filter-list {
    list-style-type: none;
    padding-left: 0px;
    overflow: auto;
    height: 100%;
}

.filter-button {
    width: 18px;
    height: 18px;
    border-radius: 2px;
    margin-left: 8px;
}

.filter-button-selected {
    background-color: var(--vscode-button-background);

    &:hover {
        background-color: var(--vscode-button-hoverBackground);
    }
}

.filter-button-selected path {
    fill: var(--vscode-button-foreground);
}
