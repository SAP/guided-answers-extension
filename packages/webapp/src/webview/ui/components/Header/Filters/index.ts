export * from './Filters';
export * from './FiltersRibbon';
