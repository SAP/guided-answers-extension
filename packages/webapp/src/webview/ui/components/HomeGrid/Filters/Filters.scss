.guided-answer__filter {
    >button {
        padding: 15px;
        width: 100%;
        display: flex;
        gap: 10px;
        flex-flow: row wrap;
        color: var(--vscode-foreground);
        background: var(--vscode-editorWidget-background);
        text-align: left;
        font-weight: 400;
        font-size: 13px;
        cursor: pointer;
        border: none;
        border-radius: 3px;

        &:hover {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
            outline-offset: -1px;
        }

        &:focus-visible {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
            outline-offset: -2px;
        }

        strong {
            font-weight: 700;
        }
    }

}
