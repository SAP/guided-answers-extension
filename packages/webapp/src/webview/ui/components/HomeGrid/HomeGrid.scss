.guided-answer__home-grid {
    display: flex;
    justify-content: center;
    gap: 5%;
    margin: 0 4%;

    &__column {
        flex: 0 1 570px;

        >div {
            margin-bottom: 30px;
        }
    }

    &__section {
        &__title {
            display: flex;
            align-items: center;
            margin: 20px 0;

            >i {
                height: 16px;
            }

            >span {
                margin: 0 5px;
            }
        }

        &__list {
            display: flex;
            flex-flow: column nowrap;
            gap: 5px;
            margin: 0;
            padding: 0;
            list-style: none;
        }
    }
}

@media only screen and (max-width: 740px) {
    .guided-answer__home-grid {
        display: block;
    }
}
