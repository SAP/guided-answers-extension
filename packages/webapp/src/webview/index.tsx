import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
