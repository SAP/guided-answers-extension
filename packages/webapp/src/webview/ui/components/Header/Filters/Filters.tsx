import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../../../types';
import { actions } from '../../../../state';
import { UiIcons } from '../../UIComponentsLib/Icons';
import { UIIconButton, UIDialog, UICheckbox, UITextInput } from '@sap-ux/ui-components';
import type { ProductFilter, ComponentFilter } from '@sap/guided-answers-extension-types';
import { Stack, IStackTokens } from '@fluentui/react';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';

import './Filters.scss';

/**
 * Sorts products array.
 *
 * @param filters A products array
 * @returns A sorted list of products
 */
export const sortProductFilters = (filters: ProductFilter[]) => {
    return filters.sort((a: ProductFilter, b: ProductFilter) => {
        return a.PRODUCT.localeCompare(b.PRODUCT, undefined, {
            numeric: true,
            sensitivity: 'base'
        });
    });
};

export const sortComponentFilters = (filters: ComponentFilter[]) => {
    type WithNoHyphens = {
        NOHYPHENS?: string;
    };
    type ComponentFilterWithNoHyphens = ComponentFilter & WithNoHyphens;

    const sorting = (a: ComponentFilterWithNoHyphens, b: ComponentFilterWithNoHyphens): any => {
        if (a.NOHYPHENS !== undefined && b.NOHYPHENS !== undefined) {
            return a.NOHYPHENS.localeCompare(b.NOHYPHENS, undefined, {
                numeric: true,
                sensitivity: 'base'
            });
        }
        return undefined;
    };

    return filters
        .map((f: ComponentFilter) => {
            let filter: ComponentFilterWithNoHyphens = f;
            filter.NOHYPHENS = f.COMPONENT.replace(/-/g, '');
            return filter;
        })
        .sort(sorting)
        .map((c: ComponentFilterWithNoHyphens): ComponentFilter => {
            delete c.NOHYPHENS;
            return c;
        });
};

/**
 *
 * @returns Filters for the header
 */
export function Filters() {
    const appState = useSelector<AppState, AppState>((state) => state);
    const PRODUCTS = 'Products';
    const COMPONENTS = 'Components';
    const [isDialogVisible, setDialogVisible] = useState(false);
    const [filter, setFilterType] = useState(PRODUCTS);
    const [productFilters, setProductFilters] = useState(appState.guidedAnswerTreeSearchResult.productFilters);
    const [componentFilters, setComponentFilters] = useState(appState.guidedAnswerTreeSearchResult.componentFilters);
    const [query, setQuery] = useState('');
    const [selectedProductFilters, setSelectedProductFilters] = useState([...appState.selectedProductFilters]);
    const [selectedComponentFilters, setSelectedComponentFilters] = useState([...appState.selectedProductFilters]);
    const isFilterProducts = filter === PRODUCTS;
    const main = {
        ['.ms-TextField-wrapper > .ms-Label']: {
            margin: 0
        },
        height: '400px'
    };

    const verticalGapStackTokens: IStackTokens = {
        childrenGap: 10,
        padding: '10px 0'
    };

    const toggleVisibility = (type: string): void => {
        if (type === PRODUCTS) {
            setFilterType(PRODUCTS);
        } else {
            setFilterType(COMPONENTS);
        }
        setQuery('');
        setDialogVisible(!isDialogVisible);
    };

    const applyFilters = (type: string): void => {
        toggleVisibility(type);
        if (type === PRODUCTS) {
            actions.setProductFilters(selectedProductFilters);
            actions.searchTree({
                query: appState.query,
                filters: {
                    product: selectedProductFilters,
                    component: selectedComponentFilters
                }
            });
        } else {
            actions.setComponentFilters(selectedComponentFilters);
            actions.searchTree({
                query: appState.query,
                filters: {
                    product: selectedProductFilters,
                    component: selectedComponentFilters
                }
            });
        }
    };

    const cancel = () => {
        setSelectedComponentFilters([...appState.selectedComponentFilters]);
        setSelectedProductFilters([...appState.selectedProductFilters]);
        setDialogVisible(false);
    };

    const toggleFilters = (type: string): void => {
        toggleVisibility(type);
        if (type === PRODUCTS) {
            setSelectedProductFilters(
                appState.guidedAnswerTreeSearchResult.productFilters
                    .filter((v) => selectedProductFilters.includes(v.PRODUCT))
                    .map((v) => v.PRODUCT)
            );
            setProductFilters(sortProductFilters(appState.guidedAnswerTreeSearchResult.productFilters));
        } else {
            setSelectedComponentFilters(
                appState.guidedAnswerTreeSearchResult.componentFilters
                    .filter((v) => selectedComponentFilters.includes(v.COMPONENT))
                    .map((v) => v.COMPONENT)
            );
            setComponentFilters(sortComponentFilters(appState.guidedAnswerTreeSearchResult.componentFilters));
        }
    };

    const resetFilter = (): void => {
        setQuery('');
        actions.searchTree({
            query: appState.query,
            filters: {
                product: appState.selectedProductFilters.length > 0 ? appState.selectedProductFilters : [],
                component: appState.selectedComponentFilters.length > 0 ? appState.selectedComponentFilters : []
            }
        });
        setDialogVisible(!isDialogVisible);
    };

    const onChange = (filter: string) => (ev?: any, checked?: boolean) => {
        if (isFilterProducts) {
            if (checked) {
                setSelectedProductFilters([...selectedProductFilters, filter]);
            } else {
                setSelectedProductFilters(selectedProductFilters.filter((e) => e !== filter));
            }
        } else if (checked) {
            setSelectedComponentFilters([...selectedComponentFilters, filter]);
        } else {
            setSelectedComponentFilters(selectedComponentFilters.filter((e) => e !== filter));
        }
    };

    useEffect(() => {
        setSelectedProductFilters(appState.selectedProductFilters);
        setSelectedComponentFilters(appState.selectedComponentFilters);
    }, [appState.selectedProductFilters, appState.selectedComponentFilters]);

    const filterType: any = {
        Products: {
            title: 'Filter Products',
            visibility: isDialogVisible,
            apply: () => applyFilters(PRODUCTS),
            listItems: productFilters.map((productFilter: { PRODUCT: string }) => (
                <UICheckbox
                    key={`${productFilter.PRODUCT}`}
                    label={productFilter.PRODUCT}
                    checked={selectedProductFilters.includes(productFilter.PRODUCT)}
                    onChange={onChange(productFilter.PRODUCT)}
                />
            ))
        },
        Components: {
            title: 'Filter Components',
            visibility: isDialogVisible,
            apply: () => applyFilters(COMPONENTS),
            listItems: componentFilters.map((componentFilter: { COMPONENT: string }) => (
                <UICheckbox
                    key={`${componentFilter.COMPONENT}`}
                    label={componentFilter.COMPONENT}
                    checked={selectedComponentFilters.includes(componentFilter.COMPONENT)}
                    onChange={onChange(componentFilter.COMPONENT)}
                />
            ))
        }
    };

    const searchFilter = (e: any) => {
        const query = e.target.value;

        if (query.length >= 1) {
            setQuery(query);
            let filters: (ProductFilter | ComponentFilter)[] =
                appState.guidedAnswerTreeSearchResult[isFilterProducts ? 'productFilters' : 'componentFilters'];

            filters = filters.filter((f: ProductFilter | ComponentFilter) =>
                isFilterProducts
                    ? (f as ProductFilter).PRODUCT.toLowerCase().includes(query.toLowerCase())
                    : (f as ComponentFilter).COMPONENT.toLowerCase().includes(query.toLowerCase())
            );

            if (isFilterProducts) {
                setProductFilters(sortProductFilters(filters as ProductFilter[]));
            } else {
                setComponentFilters(sortComponentFilters(filters as ComponentFilter[]));
            }
        } else {
            setQuery('');
            setProductFilters(sortProductFilters(appState.guidedAnswerTreeSearchResult.productFilters));
            setComponentFilters(sortComponentFilters(appState.guidedAnswerTreeSearchResult.componentFilters));
        }
    };

    return (
        <>
            <div id="filters">
                <FocusZone
                    direction={FocusZoneDirection.horizontal}
                    isCircularNavigation={true}
                    style={{ display: 'flex' }}>
                    <UIIconButton
                        id="filter-products"
                        iconProps={{ iconName: UiIcons.Table }}
                        onClick={() => toggleFilters(PRODUCTS)}
                        disabled={appState.guidedAnswerTreeSearchResult.productFilters.length === 0}
                        className={`filter-button ${selectedProductFilters.length > 0 ? 'filter-button-selected' : ''}`}
                        primary
                        title="Filter Products"></UIIconButton>
                    <UIIconButton
                        id="filter-components"
                        iconProps={{ iconName: UiIcons.IdTag }}
                        onClick={() => toggleFilters(COMPONENTS)}
                        disabled={appState.guidedAnswerTreeSearchResult.componentFilters.length === 0}
                        primary
                        title="Filter Components"
                        className={`filter-button ${
                            selectedComponentFilters.length > 0 ? 'filter-button-selected' : ''
                        }`}></UIIconButton>
                </FocusZone>
                <UIDialog
                    className="dialog-filter"
                    dialogContentProps={{ title: filterType[filter].title }}
                    isOpen={filterType[filter].visibility}
                    isBlocking={true}
                    acceptButtonText={'Apply Filter'}
                    cancelButtonText={'Cancel'}
                    styles={{ main }}
                    onAccept={() => filterType[filter].apply()}
                    onCancel={cancel}
                    onDismiss={resetFilter}>
                    <UITextInput placeholder="Search" value={query} onChange={searchFilter} />
                    <FocusZone
                        direction={FocusZoneDirection.domOrder}
                        isCircularNavigation={true}
                        style={{ overflowY: 'scroll', height: '90%', padding: 0, marginTop: '10px' }}>
                        <Stack tokens={verticalGapStackTokens} style={{ margin: '0px 4px' }}>
                            {filterType[filter].listItems}
                        </Stack>{' '}
                    </FocusZone>
                </UIDialog>
            </div>
        </>
    );
}
