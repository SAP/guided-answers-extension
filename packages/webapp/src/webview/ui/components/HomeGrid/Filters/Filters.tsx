import React from 'react';
import type { ReactElement } from 'react';
import { UIIcon, UiIcons } from '@sap-ux/ui-components';
import { useSelector } from 'react-redux';
import type { AppState } from '../../../../types';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';
import { actions } from '../../../../state';
import type { FilterStack } from '@sap/guided-answers-extension-types';
import './Filters.scss';

/**
 * Shows list of Filters.
 *
 * @returns - react elements for Filters view
 */
export function Filters(): ReactElement {
    const autoFilters = useSelector<AppState, FilterStack[]>((state) => state.autoFilters);
    const customFilters = useSelector<AppState, FilterStack[]>((state) => state.customFilters);

    const renderFilter = (f: FilterStack) => (
        <li key={`tree-item-${f.product?.join('-')}-${f.component?.join('-')}`} className="guided-answer__filter">
            <button
                onClick={(): void => {
                    actions.updateNetworkStatus('LOADING');
                    actions.searchTree({ filters: f });
                }}>
                {f.product && (
                    <div>
                        Product: <strong>{f.product.join(', ')}</strong>
                    </div>
                )}
                {f.component && (
                    <div>
                        Component: <strong>{f.component.join(', ')}</strong>
                    </div>
                )}
            </button>
        </li>
    );

    return (
        <div>
            <h3 className="guided-answer__home-grid__section__title">
                <UIIcon iconName={UiIcons.Filter} />
                <span>Filters</span>
            </h3>
            <FocusZone direction={FocusZoneDirection.vertical} isCircularNavigation={true}>
                <ul className="guided-answer__home-grid__section__list" role="listbox">
                    {autoFilters.map(renderFilter)}
                    {customFilters.map(renderFilter)}
                </ul>
            </FocusZone>
        </div>
    );
}
