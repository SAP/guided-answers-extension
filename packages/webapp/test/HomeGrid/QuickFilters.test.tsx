import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { fireEvent, screen } from '@testing-library/dom';
import { useSelector } from 'react-redux';
import { actions } from '../../src/webview/state';
import { Filters } from '../../src/webview/ui/components/HomeGrid/Filters';

jest.mock('../../src/webview/state', () => ({
    actions: {
        searchTree: jest.fn(),
        updateNetworkStatus: jest.fn()
    }
}));

jest.mock('react-redux', () => ({
    useSelector: jest.fn()
}));

describe('<Filters />', () => {
    afterEach(cleanup);

    const mockAutoFilters = [
        {
            product: ['product 1'],
            component: ['component 1']
        }
    ];

    it('Should render AutoFilters component', () => {
        (useSelector as jest.Mock).mockImplementation((selector) => selector({ autoFilters: mockAutoFilters }));
        const { container } = render(<Filters />);
        expect(container).toMatchSnapshot();

        fireEvent.click(screen.getByRole('button'));
        expect(actions.updateNetworkStatus).toBeCalledWith('LOADING');
        expect(actions.searchTree).toBeCalled();
    });
});
