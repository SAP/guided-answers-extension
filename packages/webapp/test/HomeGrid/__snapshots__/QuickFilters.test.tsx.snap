// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuickFilters /> Should render QuickFilters component 1`] = `
<div>
  <div>
    <h3
      class="guided-answer__home-grid__section__title"
    >
      <i
        aria-hidden="true"
        class="ts-icon root-105"
        data-icon-name="Filter"
      />
      <span>
        Quick Filters
      </span>
    </h3>
    <div
      class="ms-FocusZone css-109"
      data-focuszone-id="FocusZone0"
    >
      <ul
        class="guided-answer__home-grid__section__list"
        role="listbox"
      >
        <li
          class="guided-answer__quick-filter"
        >
          <button
            tabindex="0"
          >
            <div>
              Product: 
              <strong>
                product 1
              </strong>
            </div>
            <div>
              Component: 
              <strong>
                component 1
              </strong>
            </div>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
